File: CSharp/064.cs
#nullable enable

using System;
using System.Collections.Generic;

using static System.Console;
using static System.Math;
using static System.Linq.Enumerable;

using static Scanner;

static class Scanner
{
	private static string[]? s;
	private static int i;

	public static string NextToken()
	{
		if (s == null || i == s.Length) {
			s = ReadLine()!.Split();
			i = 0;
		}

		return s[i++];
	}

	public static int NextInt() => int.Parse(NextToken());
	public static long NextLong() => long.Parse(NextToken());
	public static double NextDouble() => double.Parse(NextToken());

	public static T[] ReadArray<T>(Func<string, T> selector)
		=> ReadLine()!.Split().Select(selector).ToArray();
}

public static class Program
{
	public static void Main()
	{
		int N = NextInt();
		int Q = NextInt();

		long[] A = ReadArray(long.Parse);

		long[] D = Range(0, N - 1).Select(i => A[i + 1] - A[i]).ToArray();

		long ans = D.Sum(d => Abs(d));

		foreach (int _ in Range(0, Q)) {
			int L = NextInt() - 2;
			int R = NextInt() - 1;
			long V = NextLong();

			if (L >= 0) {
				ans -= Abs(D[L]);
				D[L] += V;
				ans += Abs(D[L]);
			}

			if (R < N - 1) {
				ans -= Abs(D[R]);
				D[R] -= V;
				ans += Abs(D[R]);
			}

			WriteLine(ans);
		}
	}
}

File: Swift/064.swift
func readInt2() -> (Int, Int) {
	let s = readLine()!.split(separator: " ")
	return (Int(s[0])!, Int(s[1])!)
}

func readInt3() -> (Int, Int, Int) {
	let s = readLine()!.split(separator: " ")
	return (Int(s[0])!, Int(s[1])!, Int(s[2])!)
}

let (N, Q) = readInt2()

var A = readLine()!.split(separator: " ").map { Int($0)! }

var D = (1 ..< N).map { i in A[i] - A[i - 1] }

var ans = D.reduce(0) { acc, d in acc + abs(d) }

for _ in 0 ..< Q {
	let (L, R, V) = readInt3()
	let l = L - 2
	let r = R - 1
	
	if (l >= 0) {
		ans -= abs(D[l])
		D[l] += V
		ans += abs(D[l])
	}
	
	if (R < N) {
		ans -= abs(D[r])
		D[r] -= V
		ans += abs(D[r])
	}
	
	print(ans)
}
